### ISSUE TITLE
AnimatedMoveViewJob crash

### ISSUE THREAD

**Issue by LemonFQM** [danielgindi/Charts#3061]
AnimatedMoveViewJob don't overwrite animationEnd

**Comment by liuxuan30** [comment]
give more details please.

**Comment by liuxuan30** [comment]
I now understand this. 
```
    internal func animationEnd()
    {
        // Override this
    }
```
it changed to 
```
    internal func animationEnd()
    {
        fatalError("`animationEnd()` must be overriden by subclasses")
    }
```

**Comment by jjatie** [comment]
I think it would be better to explicitly override it with an empty implementation. Thoughts?

**Comment by liuxuan30** [comment]
yeah that's where we think differently. asking every subclass to just add an empty override is introducing duplicated code?
Using fatal what I understand is if you don't do that, bad things would happen for sure. But this is not true for the animator? As by default it does nothing at the animation end, but when you want to add something, you can add in subclass then.

**Comment by jjatie** [comment]
#3099

**Comment by liuxuan30** [comment]
I just added #3098 :)

**Comment by jjatie** [comment]
I agree that's the objective-c way to do it. In swift, `AnimatedViewPortJob` would be a protocol. This is one of those spots where we're kind of stuck between two worlds.

**Comment by liuxuan30** [comment]
when we are changing to protocol we could change then, like optional

**Comment by jjatie** [comment]
Yep! That will probably be another year or two though.

**Comment by jjatie** [comment]
@liuxuan30 by the way. How is 4.0.0 kept up to date. Would it be appropriate for me to PR master into 4.0.0?

**Comment by liuxuan30** [comment]
It's a headache, as more and more PRs coming to both master and 4.0.
one way is to let git mark the conflicts and we solve it once and for all. Or keep merging master into 4.0, and solve every conflicts. Seems both are headaches.

**Comment by jjatie** [comment]
I don't mind solving the merge conflicts every so often. I think it will be easier to do it periodically, than all at once at the end. I guess I will PR when I think the diff is becoming too big.

**Comment by liuxuan30** [comment]
ok. I will try ping daniel if this week passed.
